service/worker/workerdeployment/replay_test.go

```diff
@@ -0,0 +1,82 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package workerdeployment
+
+import (
+	"compress/gzip"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/sdk/worker"
+	"go.temporal.io/sdk/workflow"
+	deploymentspb "go.temporal.io/server/api/deployment/v1"
+	"go.temporal.io/server/common/log"
+)
+
+// TestReplays tests workflow logic backwards compatibility from previous versions.
+func TestReplays(t *testing.T) {
+	replayer := worker.NewWorkflowReplayer()
+	drainageWorkflow := func(ctx workflow.Context, args *deploymentspb.DrainageWorkflowArgs) error {
+		refreshIntervalGetter := func() any {
+			return time.Minute
+		}
+		visibilityGracePeriodGetter := func() any {
+			return time.Minute
+		}
+		return DrainageWorkflow(ctx, refreshIntervalGetter, visibilityGracePeriodGetter, args)
+	}
+	deploymentWorkflow := func(ctx workflow.Context, args *deploymentspb.WorkerDeploymentWorkflowArgs) error {
+		maxVersionsGetter := func() int {
+			return 100
+		}
+		return Workflow(ctx, maxVersionsGetter, args)
+	}
+	replayer.RegisterWorkflowWithOptions(VersionWorkflow, workflow.RegisterOptions{Name: WorkerDeploymentVersionWorkflowType})
+	replayer.RegisterWorkflowWithOptions(deploymentWorkflow, workflow.RegisterOptions{Name: WorkerDeploymentWorkflowType})
+	replayer.RegisterWorkflowWithOptions(drainageWorkflow, workflow.RegisterOptions{Name: WorkerDeploymentDrainageWorkflowType})
+
+	files, err := filepath.Glob("testdata/replay_*.json.gz")
```

Comment:
do you have a script or something you used to generate the testdata? it would be nice to have it in the repo so that if the workflow changes in the future it's easy to generate new testdata

Reply:
I was just using Antonio's sdk-go tests to generate the test data, and I feel like I don't have time to make a script to do it since this PR is on a time crunch due to the releases, but I can make a ticket to add that in the future.